### ISSUE TITLE
MMAL black screen on yuv444

### ISSUE THREAD

**Issue by oomek** [wang-bin/mdk-sdk#21]
I'm not sure if I understand this correctly. When I set multiple decoders for example
```
player.setVideoDecoders({"MMAL", "FFmpeg"});
```
does it mean that when the MMAL decoder is unable to decode one of the videos it should fall back to FFmpeg, or do I need to decide what decoder to use based on info in the file descriptor?
I'm asking because I'm getting a blank screen when playing `h264` with `yuv444` color space which is not supported by RPI4 hardware decoder. It only accepts `yuv420`.

**Comment by oomek** [comment]
I wanted to make a fallback myself. I see that the colorspace is stored in `player->mediaInfo().video[0].format` but it gives me numbers 38 or 40 and it does not correlate with the  `PixelFormat` in `VideoFrame.h` Could you tell me which enum did you use for that filed, and how to convert it to a meaningful string please?

**Comment by wang-bin** [comment]
Try the latest sdk

**Comment by oomek** [comment]
Just tried with `player->setVideoDecoders({"MMAL", "FFmpeg"});` the screen is still black for `yuv444p` unfortunately

**Comment by wang-bin** [comment]
show me your log

**Comment by oomek** [comment]
[mdk_log.txt](https://github.com/wang-bin/mdk-sdk/files/5398916/mdk_log.txt)

**Comment by wang-bin** [comment]
> yuv444p (selected)
> using avcodec software decoder...
> video stream#0 failed to decode restored packets
> [FFmpeg:h264] top block unavailable for requested intra mode -1
> 
> [FFmpeg:h264] error while decoding MB 5 0, bytestream 18435
> 
> [FFmpeg:h264] concealing 224 DC, 224 AC, 224 MV errors in I frame

What about using "mmal" decoder? It's my own implementation

**Comment by oomek** [comment]
Your implementation as I remember had more issues, but let me try again with the new lib

**Comment by oomek** [comment]
[mdk_log-mmal.txt](https://github.com/wang-bin/mdk-sdk/files/5399226/mdk_log-mmal.txt)
[mdk_log-mmal-copy2.txt](https://github.com/wang-bin/mdk-sdk/files/5399228/mdk_log-mmal-copy2.txt)

**Comment by oomek** [comment]
Maybe it's some shared lib mismatch. I'm out of ideas. I'm up do date with upgrade and firmware. I wonder what would happen if I've built mdk locally.

**Comment by wang-bin** [comment]
need to patch ffmpeg

**Comment by wang-bin** [comment]
I see errors in `mmal:copy=2` decoder. You video is 1080p 60fps. Do you have enough video memory?

**Comment by oomek** [comment]
I do not have a custom vmem defined in config as I read that RPI4 has a dynamic vram allocation:
`On the Raspberry Pi 4 the 3D component of the GPU has its own memory management unit (MMU), and does not use memory from the gpu_mem allocation. Instead memory is allocated dynamically within Linux.`

**Comment by oomek** [comment]
That was actually the case. It seems like this dynamic memory allocation does not apply to frame buffers. The 1080p videos are playing now but the performance with `mmal:copy=2` is much worse than with `MMAL` unfortunately. I still get segfaluts with any other than `copy=2` value.

**Comment by oomek** [comment]
The good news is that `yuv444p` videos are playing wirn 'mmal'

**Comment by oomek** [comment]
I'm not sure if I'm supposed to open a new issue, but I've just discovered that at the end of buffering there is a 100% CPU usage on one core, this is clearly visible for very small files like this one, when the usage is constant at 100%

Remove the txt extension:
[v_counter.mp4.txt](https://github.com/wang-bin/mdk-sdk/files/5426649/v_counter.mp4.txt)

**Comment by wang-bin** [comment]
"mmal:copy=2" should have the same performance as ffmpeg's MMAL. My patched ffmpeg can have better performance via "MMAL:copy=-1". "mmal:copy=1" should has the same performance as "MMAL:copy=-1", but now the bug is it paused at the first frame.

**Comment by oomek** [comment]
Out of curiosity, do you `sws_scale` or draw in native format with shader based color space conversion?

**Comment by wang-bin** [comment]
always use shaders

**Comment by oomek** [comment]
Just to let you know, this issue still remains on 0.10.2

**Comment by wang-bin** [comment]
It's a regression since 5 months ago(unable to switch to a new decoder). Should be fixed in the latest sdk.
